gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -421,13 +478,24 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) {
   }
 
   /**
-   * This feature is not supported. Please use {@link #readAttributes(Path, Class, LinkOption...)}
-   *
-   * @throws UnsupportedOperationException
    */
   @Override
   public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) {
-    throw new UnsupportedOperationException();
+    CloudStoragePath cloudPath = checkPath(path);
+    checkNotNull(attributes);
+    BlobInfo blobInfo = gcsStorage.get(cloudPath.getBlobId());
+    HashMap<String, Object> ret = new HashMap<>();
+    Map<String, String> metadata = blobInfo.metadata();
+    // it's faster to copy everything than to parse the attributes string.
+    if (null != metadata) {
+      for (Map.Entry<String, String> entry : metadata.entrySet()) {
+        ret.put(entry.getKey(), entry.getValue());
+      }
+    }
+    ret.put("content-type", blobInfo.contentType());
+    ret.put("cache-control", blobInfo.cacheControl());
+    ret.put("generation", blobInfo.generation());
+    return ret;
```

Comment:
Removed the implementation for this method and fixed the tests so they don't need it.